jsk_arc2017_baxter/euslisp/lib/apc-interface.l

```diff
@@ -0,0 +1,380 @@
+;; -*- mode: lisp;-*-
+
+(require "package://jsk_2016_01_baxter_apc/euslisp/jsk_2016_01_baxter_apc/baxter-interface.l")
+(require "package://jsk_2015_05_baxter_apc/euslisp/jsk_2015_05_baxter_apc/util.l")
+(require "package://jsk_2016_01_baxter_apc/euslisp/jsk_2016_01_baxter_apc/baxter.l")
+(require "package://jsk_2016_01_baxter_apc/euslisp/jsk_2016_01_baxter_apc/util.l")
+(require "package://jsk_arc2017_baxter/euslisp/lib/util.l")
+
+(ros::load-ros-manifest "jsk_arc2017_baxter")
+
+(unless (find-package "JSK_ARC2017_BAXTER")
+  (make-package "JSK_ARC2017_BAXTER"))
+
+(defclass jsk_arc2017_baxter::transform-listener
+  :super ros::transform-listener)
+
+(defmethod jsk_arc2017_baxter::transform-listener
+  (:tf-pose->coords
+    (frame_id pose)
+    (send (send self :lookup-transform "base" frame_id (ros::time 0))
+          :transform (ros::tf-pose->coords pose))))
+
+(defclass jsk_arc2017_baxter::apc-interface
+  :super jsk_2016_01_baxter_apc::baxter-interface
+  :slots (object-boxes-in-bin-
+          object-coords-in-bin-
+          bin-movable-regions-))
+
+(defmethod jsk_arc2017_baxter::apc-interface
+  (:init
+    (&rest args &key
+           ((:moveit-environment mvit-env)
+            (instance jsk_2016_01_baxter_apc::baxter-moveit-environment))
+           ((:moveit-robot mvit-rb) (instance jsk_2016_01_baxter_apc::baxter-robot :init))
+           &allow-other-keys)
+    (send-super* :init :moveit-environment mvit-env
+                 :moveit-robot mvit-rb args)
+    ;; initialize slots
+    (setq tfl- (instance jsk_arc2017_baxter::transform-listener :init))
+    (setq object-boxes-in-bin- (make-hash-table))
+    (setq bin-movable-regions- (make-hash-table))
+    (setq object-boxes-in-bin- (make-hash-table))
+    (setq object-coords-in-bin- (make-hash-table)))
+  (:bbox->cube
+    (bbox)
+    (let* ((dims (ros::tf-point->pos (send bbox :dimensions)))
+           (bx (make-cube (elt dims 0) (elt dims 1) (elt dims 2))))
+      (send bx :newcoords
+            (send tfl- :tf-pose->coords
+                  (send bbox :header :frame_id)
+                  (send bbox :pose)))
+      bx))
+  (:cube->movable-region
+    (cube &key (offset (list 0 0 0)))
+    (let (movable-region vertex-values)
+      (send cube :worldcoords)
+      (dotimes (i 3)
+        (setq vertex-values (mapcar #'(lambda (x) (aref x i)) (send cube :vertices)))
+        (pushback (list
+                    (+ (apply #'min vertex-values) (elt offset i))
+                    (- (apply #'max vertex-values) (elt offset i)))
+                  movable-region))
+      movable-region))
+  (:get-movable-region-for-bin
```

Comment:
In my thought, this is not *get* function.
This is *set* function for `bin-movable-regions-`